maximize-tensorflow-performance-on-cpu-considerations-and-recommendations-for-inference/maximize-tensorflow-performance-on-cpu-considerations-and-recommendations-for-inference.md

```diff
@@ -0,0 +1,287 @@
+  <br><br>
+  <li>intra_op_parallelism_threads</li>
+  <br>
+  This runtime setting controls parallelism inside an operation. For instance, if matrix multiplication or reduction is intended to be executed in several threads, this variable should be set. TensorFlow will schedule tasks in a thread pool that contains intra_op_parallelism_threads threads. As illustrated later in Figure 2, OpenMP* threads are bound to thread context as close as possible on different cores. Setting this environment variable to the number of available physical cores is recommended.
+  <br><br>
+  <li>inter_op_parallelism_threads</li>
+  <br>
+  NOTE: This setting is highly dependent on hardware and topologies, so it’s best to empirically confirm the best setting on your workload.
+  <br><br>
+  This runtime setting controls parallelism among independent operations. Since these operations are not relevant to each other, TensorFlow will try to run them concurrently in the thread pool that contains inter_op_parallelism_threads threads. This variable should be set to the number of parallel paths where you want the code to run. For Intel® Optimization for TensorFlow, we recommend starting with the setting '2’, and adjusting after empirical testing.
+</details>
+
+<details>
+  <summary>Data layout</summary>
+  <br>
+  <b>Recommended settings → data_format = NHWC</b>
+<br>
+tf_cnn_benchmarks usage (shell)
+<br>
+<pre>python tf_cnn_benchmarks.py --num_intra_threads=&lt;number of physical cores per socket&gt; --num_inter_threads=&lt;number of sockets&gt; --data_format=NHWC</pre>
+<br>
+Efficiently using cache and memory yields remarkable improvements in overall performance. A good memory access pattern minimizes extra cost for accessing data in memory and improves overall processing. Data layout, how data is stored and accessed, plays an important role in achieving these good memory access patterns. Data layout describes how multidimensional arrays are stored linearly in memory address space.
+
+In most cases, data layout is represented by four letters for a two-dimensional image:
+
+- N: Batch size, indicates number of images in a batch.
+- C: Channel, indicates number of channels in an image.
+- W: Width, indicates number of horizontal pixels of an image.
+- H: Height, indicates number of vertical pixels of an image.
+<br>
+The order of these four letters indicates how pixel data are stored in the one-dimensional memory space. For instance, NCHW indicates pixel data are stored as width first, then height, then channel, and finally batch (Illustrated in Figure 2). The data is then accessed from left-to-right with channel-first indexing. NCHW is the recommended data layout for using oneDNN, since this format is an efficient data layout for the CPU. TensorFlow uses NHWC as its default data layout, but it also supports NCHW.
+
+![Data Formats for Deep Learning NHWC and NCHW](https://www.intel.com/content/dam/develop/external/us/en/images/data-layout-nchw-nhwc-804042.png) 
```

Comment:
this should replace with link of AEM graph /content/dam/www/central-libraries/us/en/images/data-layout-nchw-nhwc-804042.png

Reply:
Updated